File: output_20260318_123118/deepseek/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

struct person {
  char* name;
  struct person* spouse;
};

/*@
predicate person(struct person* p;) =
  p->name |-> _ &*& p->spouse |-> ?sp &*& malloc_block_person(p);
@*/

/*@
predicate married(struct person* p, struct person* q;) =
  p->spouse |-> q &*& q->spouse |-> p &*& malloc_block_person(p) &*& malloc_block_person(q);
@*/

/*@
predicate single(struct person* p;) =
  p->spouse |-> 0 &*& malloc_block_person(p);
@*/

struct person *create_person()
  //@ requires true;
  //@ ensures single(result);
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  p->spouse = 0;
  //@ close single(p);
  return p;
}

void marry(struct person *this, struct person *other)
  //@ requires single(this) &*& single(other);
  //@ ensures married(this, other);
{
  //@ open single(this);
  //@ open single(other);
  this->spouse = other;
  other->spouse = this;
  //@ close married(this, other);
}

struct person* person_get_spouse(struct person* this)
  //@ requires person(this);
  //@ ensures person(this) &*& result == this->spouse;
{
  //@ open person(this);
  struct person* sp = this->spouse;
  //@ close person(this);
  return sp;
}

void divorce(struct person* this)
  //@ requires married(this, ?other);
  //@ ensures single(this) &*& single(other);
{
  //@ open married(this, other);
  this->spouse->spouse = 0;
  this->spouse = 0;
  //@ close single(this);
  //@ close single(other);
}

void die(struct person *this)
  //@ requires person(this);
  //@ ensures true;
{
  //@ open person(this);
  if(this->spouse != 0) {
    //@ open person(this->spouse);
    this->spouse->spouse = 0;
    //@ close person(this->spouse);
  }
  free(this);
}

int main() 
  //@ requires true;
  //@ ensures true;
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}

File: output_20260319_145930/deepseek/java/Linklist/Spouse.java
class Person {

    protected Person spouse;
    
    //@ predicate Person() = spouse == null;
    //@ predicate Married(Person other) = spouse == other &*& other != null &*& other.spouse == this;

    public  void spouse_symm()
        //@ requires Person();
        //@ ensures Person();
    {
        //@ close Person();
    }

    public Person()
        //@ ensures Person();
    {
        //@ close Person();
    }
    
    public Person getSpouse()
        //@ requires Person();
        //@ ensures Person() &*& result == null;
    {
        //@ open Person();
        return spouse;
        //@ close Person();
    }
    
    protected void setSpouse(Person other)
        //@ requires Person() &*& other != null &*& other.Person();
        //@ ensures Married(other);
    {
        //@ open Person();
        //@ open other.Person();
        spouse = other;
        other.spouse = this;
        //@ close Married(other);
    }
    
    protected void clearSpouse()
        //@ requires Married(?other);
        //@ ensures Person() &*& other.Person();
    {
        //@ open Married(?other);
        spouse.spouse = null;
        spouse = null;
        //@ close Person();
        //@ close other.Person();
    }
    
    void marry(Person other)
        //@ requires Person() &*& other != null &*& other.Person();
        //@ ensures Married(other);
    {
        other.setSpouse(this);
    }
    
    void divorce()
        //@ requires Married(?other);
        //@ ensures Person() &*& other.Person();
    {
        spouse.clearSpouse();
    }

}

class Program {

    public static void main(String[] args)
        //@ requires true;
        //@ ensures true;
    {
        Person a = new Person();
        Person b = new Person();
        a.marry(b);
        b.divorce();
    }

}